### ISSUE TITLE
SQL Query rewriting logic using TVs

### ISSUE THREAD

**Issue by bazooka720** [DataJunction/dj#1068]
One of the features of semantic layer like cube.js has is that it has ability to rewrite query to identify whether to use pre-age (MV) tables vs. go to source or some combination by studying the query dimensions. 

Curious, if such a feature exists or planned? In addition, a feature to use history to identify new TVs to create automatically would be super helpful as well

**Comment by bazooka720** [comment]
Yes.  Basically if DJ finds a MV (materialized view based on query
pattern), database query is rewritten to become a multi-stage process: DJ
checks if an up-to-date copy of the MV exists.DJ will then execute a query
against the MV tables instead of the raw data. If the query requires / can
be fulfilled partly from materialized and partly from source, it can do it
as well (Federated as Trino Supports it).

In general, can you point to various functionality that DJ has as it
relates to query rewrite and automation relating to it?  In the doc you
indicate following functionality but would appreciate more details:

Discover New Dimensions️
Find the metrics you’re interested in and instantly discover the dimensions
that are common among them.

Understand Dependencies
DJ understands the dependencies for metrics across the entire system using
a robust metadata layer.

Prevent Broken Metrics
Develop quickly and let DJ highlight the impact each change will have on
all metrics.

Thanks!

**Comment by shangyian** [comment]
> Yes.  Basically if DJ finds a MV (materialized view based on query
pattern), database query is rewritten to become a multi-stage process: DJ
checks if an up-to-date copy of the MV exists.DJ will then execute a query
against the MV tables instead of the raw data.

Yep, our query generation process does exaclty this.

> If the query requires / can
be fulfilled partly from materialized and partly from source, it can do it
as well (Federated as Trino Supports it).

We don't do this part just yet.

> In general, can you point to various functionality that DJ has as it
relates to query rewrite and automation relating to it?  In the doc you
indicate following functionality but would appreciate more details:

Yeah, we need to update the docs, but we've been in a rapid development phase and haven't been as focused on the docs. Will get to this in the next few weeks!
